File: src/order/infrastructure/FileOrderRepository.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { Order, OrderProps } from "../domain/Order.js";
import { OrderLineItem } from "../domain/OrderLineItem.js";
import { OrderRepository } from "../domain/OrderRepository.js";

/**
 * FileOrderRepository — Infrastructure Implementation
 *
 * Persists Order aggregates to a local JSON file.
 * Implements the OrderRepository interface defined in the domain layer.
 */

interface StoredOrder {
  orderId: string;
  customerId: string;
  status: "DRAFT";
  items: {
    productId: string;
    quantity: number;
    unitPrice: number;
    subtotal: number;
  }[];
  total: number;
  createdAt: string;
}

interface StoreData {
  orders: StoredOrder[];
  nextSequence: number;
}

export class FileOrderRepository implements OrderRepository {
  private readonly filePath: string;

  constructor(filePath: string) {
    this.filePath = filePath;
  }

  async save(order: Order): Promise<void> {
    const data = this.readStore();
    data.orders.push(order.toJSON() as unknown as StoredOrder);
    this.writeStore(data);
  }

  async findById(orderId: string): Promise<Order | null> {
    const data = this.readStore();
    const stored = data.orders.find((o) => o.orderId === orderId);
    if (!stored) return null;
    return this.toDomain(stored);
  }

  async findAll(): Promise<Order[]> {
    const data = this.readStore();
    return data.orders.map((stored) => this.toDomain(stored));
  }

  async nextId(): Promise<string> {
    const data = this.readStore();
    const seq = data.nextSequence;
    data.nextSequence = seq + 1;
    this.writeStore(data);
    return `ORD-${String(seq).padStart(4, "0")}`;
  }

  private toDomain(stored: StoredOrder): Order {
    const items = stored.items.map((i) =>
      OrderLineItem.fromData({
        productId: i.productId,
        quantity: i.quantity,
        unitPrice: i.unitPrice,
        subtotal: i.subtotal,
      })
    );

    const props: OrderProps = {
      orderId: stored.orderId,
      customerId: stored.customerId,
      status: stored.status,
      items,
      total: stored.total,
      createdAt: stored.createdAt,
    };

    return Order.fromData(props);
  }

  private readStore(): StoreData {
    if (!existsSync(this.filePath)) {
      return { orders: [], nextSequence: 1 };
    }
    try {
      const raw = readFileSync(this.filePath, "utf-8");
      return JSON.parse(raw) as StoreData;
    } catch {
      return { orders: [], nextSequence: 1 };
    }
  }

  private writeStore(data: StoreData): void {
    const dir = dirname(this.filePath);
    if (!existsSync(dir)) {
      mkdirSync(dir, { recursive: true });
    }
    writeFileSync(this.filePath, JSON.stringify(data, null, 2), "utf-8");
  }
}

File: src/order/infrastructure/file_order_repository.py
"""FileOrderRepository — JSON-file-backed implementation of OrderRepository."""

from __future__ import annotations

import json
import os
from pathlib import Path
from typing import List, Optional

from src.order.domain.order import Order
from src.order.domain.order_repository import OrderRepository


class FileOrderRepository(OrderRepository):
    """Persist orders to a local JSON file so they survive restarts.

    Default storage location: ``~/.oms-5k/orders.json``
    """

    def __init__(self, storage_path: str | None = None) -> None:
        if storage_path is None:
            storage_dir = Path.home() / ".oms-5k"
        else:
            storage_dir = Path(storage_path).parent
        storage_dir.mkdir(parents=True, exist_ok=True)

        self._path = (
            Path(storage_path)
            if storage_path
            else storage_dir / "orders.json"
        )
        self._orders: List[Order] = self._load()

    # ------------------------------------------------------------------
    # OrderRepository interface
    # ------------------------------------------------------------------
    def next_id(self) -> str:
        """Return the next sequential ID like ORD-0001."""
        seq = len(self._orders) + 1
        return f"ORD-{seq:04d}"

    def save(self, order: Order) -> None:
        self._orders.append(order)
        self._flush()

    def find_by_id(self, order_id: str) -> Optional[Order]:
        for order in self._orders:
            if order.order_id == order_id:
                return order
        return None

    def find_all(self) -> List[Order]:
        return list(self._orders)

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------
    def _load(self) -> List[Order]:
        if not self._path.exists():
            return []
        try:
            with open(self._path, "r", encoding="utf-8") as fh:
                data = json.load(fh)
            return [Order.from_dict(d) for d in data]
        except (json.JSONDecodeError, KeyError):
            return []

    def _flush(self) -> None:
        with open(self._path, "w", encoding="utf-8") as fh:
            json.dump(
                [o.to_dict() for o in self._orders],
                fh,
                indent=2,
            )
